text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+
+      ```rust=
+      fn do_stuff<T: Add>() -> usize { std::mem::size_of::<<T as Add>::Output>() }
+      //~^ error! the trait bound `<T as Add>::Output: Sized` is not satisfied
+      ```
+
+      Relaxing the bounds of an associated type is in effect giving existing
+      parameters a less restrictive bound which is not backwards compatible.
+[^5]: If it was desirable to add a new trait to this RFC's proposed hierarchy
+      then there are four possibilities:
+
+      - Before `Sized`
+        - i.e. `NewSized: Sized: ValueSized: Pointee`
+        - Adding bounds on a new trait before `Sized` is not
+          backwards-compatible as described above in [Implementing
+          `Sized`][implementing-sized].
+      - Between `Sized` and `ValueSized`
+        - i.e. `Sized: NewSized: ValueSized: Pointee`
+        - Adding bounds on a new trait between `Sized` and `ValueSized` is
+          backwards compatible for some functions (except for on a trait's
+          associated type). An existing function with a `Sized` bound could
+          be relaxed to the new trait without breaking existing code.
+          Existing functions with a `ValueSized` or `Pointee` bound could not.
+      - Between `ValueSized` and `Pointee`
+        - i.e. `Sized: ValueSized: NewSized: Pointee`
+        - Adding bounds on a new trait between `ValueSized` and `Pointee` is
+          backwards compatible for some functions (except for on a trait's
+          associated type). An existing function with a `Sized` or `ValueSized`
+          bound could be relaxed to the new trait without breaking existing code.
+          Existing functions with a `Pointee` bound could not.
+      - After `Pointee`
+        - i.e. `Sized: ValueSized: Pointee: NewSized`
+        - For the same reasons as the traits proposed in this RFC, adding bounds
+          on a new trait after `Pointee` is backwards compatible (except for on a
+          trait's associated type). Any existing function will have stricter
+          bounds than the new trait and so could be relaxed to the new trait
+          without breaking existing code.
+
+## `Sized` bounds
+[sized-bounds]: #sized-bounds
+
+`?Sized` would be made syntactic sugar for a `const ValueSized` bound. A
+`const ValueSized` bound is equivalent to a `?Sized` bound as all values in Rust
+today whose types do not implement `Sized` are valid arguments to
+[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
+computed given a value and knowledge of the type and target platform, and
+therefore will implement `const ValueSized`. As there are currently no
+extern types or other types which would not implement `const ValueSized`,
+every type in Rust today which would satisfy a `?Sized` bound would satisfy
+a `const ValueSized` bound.
+
+**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
+a `const ValueSized` bound. In the next edition, use of `?Sized` syntax will be prohibited
```

Comment:
> `?Sized` syntax will be prohibited

What about `?Trait` (`Trait` ≠ `Sized`). Feels a bit weird to reject `?Sized` in Rust N and continue to permit all other (no-op) `?Trait` bounds.